### PR TITLE
Add QueryMap class and encoder

### PR DESCRIPTION
**Pull request by nmiyake**
Allow Feign clients to use a QueryMap to send GET
requests that have arbitrary query parameter keys
and values. Analogous to the functionality provided
by the @QueryMap annotation in Retrofit.

### CODE REVIEW

**Comment by nmiyake** [comment]
@uschi2000 @markelliot 

Not dead-set on getting this merged as part of this repo (could just have this code in my client project as well), but interested in getting your feedback on the general concept/approach.

Our current Retrofit client uses the `@QueryMap` functionality in Retrofit, and this is an attempt to support the same concept in Feign (approach is to define a new object that's passed to a GET request as a `@Param` object, and a custom encoder reads the map provided by the object and embeds them as query params in the request).

Rob, this implements some of the stuff we talked about. Although this makes the most sense for client code, it does support the server scenario pretty naturally as well (see unit test where there's a strongly typed interface, but there can be a "Client" interface that extends it to provide QueryMap support)

**Comment by uschi2000** [comment]
I see two use-cases for this, correct if I'm wrong:
- a method with a large (yet finite) number of optional (yet statically known) parameters, defined by a server interface with hard-coded parameters
- a method with an unbounded (and not statically known) number of parameters

**Comment by nmiyake** [comment]
Correct -- there may be others, but those are the main cases that come to mind for me as well.

File: http-clients/src/main/java/com/palantir/remoting/http/QueryMap.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import com.google.common.collect.Multimap;
+import org.immutables.value.Value;
+import org.immutables.value.Value.Style.ImplementationVisibility;
+
+@JsonDeserialize(as = ImmutableQueryMap.class)
+@JsonSerialize(as = ImmutableQueryMap.class)
+@Value.Immutable
+@Value.Style(visibility = ImplementationVisibility.PACKAGE)
+public abstract class QueryMap {
+    public abstract Multimap<String, String> queryMap();
```
Comment (markelliot) [inline http-clients/src/main/java/com/palantir/remoting/http/QueryMap.java:14]: do Multimaps actually serialize reasonably?

File: http-clients/src/test/java/feign/QueryMapUsingServerWithContextTests.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+    @Before
+    public void before() {
+        String endpointUri = "http://localhost:" + APP.getLocalPort();
+
+        FeignClientFactory factory = FeignClientFactory.of(
+                new JAXRSContract(),
+                // register query map encoder
+                new QueryMapEncoder(new JacksonEncoder(ObjectMappers.guavaJdk7())),
+                new OptionalAwareDecoder(new TextDelegateDecoder(new JacksonDecoder(ObjectMappers.guavaJdk7()))),
```
Comment (markelliot) [inline http-clients/src/test/java/feign/QueryMapUsingServerWithContextTests.java:50]: we really need to pay down the debt on this at some point

**Comment by markelliot** [comment]
thoughts on contributing this directly to Feign?

File: http-clients/src/main/java/com/palantir/remoting/http/QueryMap.java (hunk 1 above)
Comment (nmiyake) [inline http-clients/src/main/java/com/palantir/remoting/http/QueryMap.java:14]: Yes -- just tried it in a unit test, and Multimaps can be written to JSON and read back from JSON properly (at least when using the JDK 7 Guava ObjectMapper).

However, after thinking about it, this object doesn't actually need to support JSON serialization since it's converted to query parameters on the client side and is never actually passed over the wire. Will remove the annotations.

File: http-clients/src/test/java/feign/QueryMapUsingServerWithContextTests.java (hunk 2 above)
Comment (nmiyake) [inline http-clients/src/test/java/feign/QueryMapUsingServerWithContextTests.java:50]: Can you clarify what debt you're referring to?

**Comment by nmiyake** [comment]
Thought about it, but don't know if it would fit their model exactly -- poked around their source code, and from what I can tell they don't ship with any custom `Encoder` subclasses. Although I haven't spent a ton of time on this, I did look through their codebase a bit to see if there would be a cleaner way to add it there, but couldn't come up with anything that would be straight-forward.

My calculus was basically that this solution was just simple/flexible/clean enough for me to be comfortable to use as a client, but not quite elegant enough build directly into Feign (at least in its current form). However, if either of you have thoughts on how this might be done in Feign directly, please let me know and would be open to looking into it.

File: http-clients/src/test/java/feign/QueryMapUsingServerWithContextTests.java (hunk 2 above)
Comment (markelliot) [inline http-clients/src/test/java/feign/QueryMapUsingServerWithContextTests.java:50]: Sorry meant to tag Rob with that. We have a growing number of encoders and decoders and should really come up with a different config pattern for adding all of them. Feel free to ignore this.

**Comment by markelliot** [comment]
Would you mind opening an issue and possibly linking to this PR as what it might look like on the Feign repo? They've been super responsive and it'd be cool to float the idea.

**Comment by nmiyake** [comment]
Sure, will do.

File: http-clients/src/test/java/feign/QueryMapUsingServerWithContextTests.java (hunk 2 above)
Comment (nmiyake) [inline http-clients/src/test/java/feign/QueryMapUsingServerWithContextTests.java:50]: OK cool, yeah, that makes sense. Tagging @uschi2000 and throwing out some ideas that came to my mind (noticed this as well when I was writing this code).

If most of the decoders/encoders are delegating ones, could have a pattern where you just supply them all and have them chained. Should work for most simple cases, although in some cases could lead to some weird dependencies. 

Another approach could be to have some kind of "handler" pattern where the thing you're registering has a predicate to determine whether or not to process the item and will then process it if so. In this scenario, builder/config could take a collection of processors and have a contract that any input will be processed by the first processor whose predicate is satisfied (although this can have corner cases depending on the situation as well like if a handler for a compound type wants to use the "Optional" handler for the sub-components of its type).

Comment (uschi2000) [inline http-clients/src/test/java/feign/QueryMapUsingServerWithContextTests.java:50]: +1, that's pretty much exactly what I had suggested before.

**Comment by nmiyake** [comment]
Closing this out since I think this will use a different approach now.

My PR for adding `@QueryMap` to Feign is progressing, and should hopefully merge soon. This should allow it to be supported client-side natively.

There will still be a bit of work since this annotation will only be on `Contract.Default` and not on `JAXRSContract`, which is what we use for the builders in http-remoting. I may need to implement something like `JAXRSContractWithQueryParams` to add support for the annotation to JAXRS and then have http-remoting use that or something (or make the builders more permissive). Will submit a PR for that appraoch once I have it.
